alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/FlowFieldSteeringAction.kt

```diff
@@ -2,45 +2,68 @@ package it.unibo.alchemist.model.implementations.actions
+import java.lang.IllegalStateException
 
 /**
  * Generic implementation of an action influenced by the concentration of a given molecule in the environment.
  *
  * @param env
  *          the environment inside which the pedestrian moves.
+ * @param reaction
+ *          the reaction which executes this action.
  * @param pedestrian
  *          the owner of this action.
  * @param targetMolecule
  *          the {@link Molecule} you want to know the concentration in the different positions of the environment.
- * @param selectPosition
- *          the logic according to the target position is determined from all the positions checked.
  */
-open class FlowFieldSteeringAction<T>(
-    private val env: Physics2DEnvironment<T>,
-    reaction: Reaction<T>,
-    private val pedestrian: Pedestrian2D<T>,
-    private val targetMolecule: Molecule,
-    private val selectPosition: Iterable<Euclidean2DPosition>.(Molecule) -> Euclidean2DPosition,
-    targetSelectionStrategy: TargetSelectionStrategy<Euclidean2DPosition>
-) : SteeringActionImpl<T, Euclidean2DPosition>(
-    env,
-    reaction,
-    pedestrian,
-    targetSelectionStrategy
-) {
+abstract class FlowFieldSteeringAction(
+    protected val env: Physics2DEnvironment<Number>,
+    reaction: Reaction<Number>,
+    pedestrian: Pedestrian2D<Number>,
+    protected val targetMolecule: Molecule
+) : SteeringActionImpl<Number, Euclidean2DPosition>(env, reaction, pedestrian) {
 
-    override fun interpolatePositions(
-        current: Euclidean2DPosition,
-        target: Euclidean2DPosition,
-        maxWalk: Double
-    ): Euclidean2DPosition =
-        current.surrounding(env, maxWalk)
-                .filter { env.canNodeFitPosition(pedestrian, it) }
-                .toMutableList()
-                .selectPosition(targetMolecule) - current
+    override fun nextPosition(): Euclidean2DPosition = env.getLayer(targetMolecule)
+        .orElseThrow { IllegalStateException("no layer containing $targetMolecule could be found") }
+        .let { layer ->
+            currentPosition.surrounding(env, maxWalk())
+                /*
+                 * Next relative position.
+                 */
+                .selectPosition(layer, layer.concentrationIn(currentPosition)) - currentPosition
+        }
+
+    /**
+     * Selects a position from a list, given the layer and its concentration in the current position
+     * of the pedestrian.
+     * This function contains the policy concerning how to select the most desirable position.
+     */
+    abstract fun List<Euclidean2DPosition>.selectPosition(
+        layer: Layer<Number, Euclidean2DPosition>,
+        currentConcentration: Double
+    ): Euclidean2DPosition
+
+    /**
+     * Discard the positions the node can't fit.
+     */
+    protected fun MutableList<Euclidean2DPosition>.discardUnsuitablePositions(): List<Euclidean2DPosition> =
+        map {
+            if (env is EuclideanPhysics2DEnvironmentWithObstacles<*, Number>) {
```

Comment:
Uhmm... don't we have a superinterface for environments with obstacles? we just need `next`
An the if may become a one-liner

Reply:
With "one-liner" you mean `if .. else ..` on a single line or `map { if ..`?

Reply:
possibly the latter
If the type name is short, you can
`map { if (env is X<A, B>) somemethod() else null }`
going to a single new line in a `map` rarely works, you gain a single column due to indentation.

Reply:
Ok, I moved the if statement and its condition on the same line of `map`. Branches are too long to be contained in a single line.